windows-agent/internal/daemon/networking.go

```diff
@@ -140,14 +141,15 @@ func getAddrList(opts options) (head *ipAdapterAddresses, err error) {
 	var buff buffer[ipAdapterAddresses]
 
 	// Win32 API docs recommend a buff size of 15KB.
-	buff.resizeBytes(15 * kilobyte)
-
+	size := 15 * kilobyte
 	for range 10 {
-		size := buff.byteCount()
-		err := opts.getAdaptersAddresses(family, flags, 0, &buff.data[0], &size)
+		size, err = buff.resizeBytes(size)
+		if err != nil {
+			return nil, err
+		}
+		err = opts.getAdaptersAddresses(family, flags, 0, &buff.data[0], &size)
 		if errors.Is(err, ERROR_BUFFER_OVERFLOW) {
 			// Buffer too small, try again with the returned size.
-			buff.resizeBytes(size)
```

Comment:
I made resizing the buffer to fail if the resulting allocation would exceed MaxUint32, so we enforce the integer arithmetic safety. At the first glance it seems impossible, but if we pass a value close to MaxUint32 and that's not a multiple of the size of the type T, than we increment the amount of T's we'd allocate, which may lead to more than MaxUint32 bytes (that's why I'm doing my math with uint64 in the first place). We still need the size to be uint32 because that's what `GetAdapterAddresses` works with.

I'm placing the resize call back to where they were. Having it inside the loop would allow checking for its error only once.

The loop only exists for one reason: cope with the inherently racy Win32 GetAdapterAddresses syscall. It's inherently racy because:
- the caller must provide enough memory for it to write the IP adapter addresses table (15 kB in our case)
- the memory supplied by the caller might be insuficient;
- it advises the amount of memory needed when that happens;
- the caller then needs to allocate more memory;
- the table might have changed enough so the new buffer is not enough either.

It seems rather impractical condition, but that's the recommended way to call GetAdapterAddresses ( [see the MS docs](https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses#:~:text=The%20recommended%20method,function%20multiple%20times.) )
That pattern of double-syscall is quite common on the Windows world: call me once so I tell you my preconditions so you can call me the second time the right way :) They even have an open source C++ library that handles this kind of call generically and automatically (among other things :) )

![image](https://github.com/user-attachments/assets/96dc2a35-5936-459d-8afc-bef68ec728fb)